pkg/services/ngalert/schedule/registry.go

```diff
@@ -233,3 +233,121 @@ func (r *alertRulesRegistry) getDiff(rules map[models.AlertRuleKey]*models.Alert
 	}
 	return result
 }
+
+type fingerprint uint64
+
+func (f fingerprint) String() string {
+	return fmt.Sprintf("%016x", uint64(f))
+}
+
+// fingerprintSeparator used during calculation of fingerprint to separate different fields. Contains a byte sequence that cannot happen in UTF-8 strings.
+var fingerprintSeparator = []byte{255}
+
+type ruleWithFolder struct {
+	rule        *models.AlertRule
+	folderTitle string
+}
+
+// fingerprint calculates a fingerprint that includes all fields except rule's Version and Update timestamp.
+func (r ruleWithFolder) Fingerprint() fingerprint {
+	rule := r.rule
+
+	sum := fnv.New64()
+
+	writeBytes := func(b []byte) {
+		_, _ = sum.Write(b)
+		_, _ = sum.Write(fingerprintSeparator)
+	}
+	writeString := func(s string) {
+		// avoid allocation when converting string to byte slice
+		writeBytes(*(*[]byte)(unsafe.Pointer(&s))) //nolint:gosec
+	}
+	// this temp slice is used to convert ints to bytes.
+	tmp := make([]byte, 8)
+	writeInt := func(u int64) {
+		binary.LittleEndian.PutUint64(tmp, uint64(u))
+		writeBytes(tmp)
+	}
+
+	// allocate a slice that will be used for sorting keys, so we allocate it only once
+	var keys []string
+	maxLen := int(math.Max(math.Max(float64(len(rule.Annotations)), float64(len(rule.Labels))), float64(len(rule.Data))))
+	if maxLen > 0 {
+		keys = make([]string, maxLen)
+	}
+
+	writeLabels := func(lbls map[string]string) {
+		// maps do not guarantee predictable sequence of keys.
+		// Therefore, to make hash stable, we need to sort keys
+		if len(lbls) == 0 {
+			return
+		}
+		idx := 0
+		for labelName := range lbls {
+			keys[idx] = labelName
+			idx++
+		}
+		sub := keys[:idx]
+		sort.Strings(sub)
+		for _, name := range sub {
+			writeString(name)
+			writeString(lbls[name])
+		}
+	}
+	writeQuery := func() {
+		// The order of queries is not important as they represent an expression tree.
+		// Therefore, the order of elements should not change the hash. Sort by RefID because it is the unique key.
+		for i, q := range rule.Data {
+			keys[i] = q.RefID
```

Comment:
Won't this overwrite data from `writeLabels`?

Reply:
Yes, slice `keys` is  used for reading only right after mutation and sorting. a function that uses it exists the information in it is not needed anymore and is overwritten by the next user (i.e. another writeLabels or writeQuery)

Reply:
no need because the next usage will just overwrite first n-elements it needs and will take a subslice